### PR TITLE
Add wildcard to inject directive

### LINKED ISSUE #1206: Add wildcard path for @inject directive
Considering this example schema

`
type User {
    id: ID!
    name: String!
    comments: [Comment!]! @hasMany
}

type Post {
    id: ID!
    title: String!
    body: String!
    user: User! @belongsTo
}

type Comment {
    id: ID!
    title: String!
    body: String!
    user: User! @belongsTo
}

input UpdateUserInput {
    comments: UpdateCommentsHasMany
}

input UpdateCommentsHasMany {
    create: [CreateCommentInput!]
}

input CreateCommentInput {
    post_id: ID!
    title: String!
    body: String!
}

type Mutation @guard {
    updateUser(input: UpdateUserInput!): User 
        @update
        @inject(context: "user.id", name: "comments.create.*.user_id")
}
`

It would be awesome if Lighthouse could inject data in the argument set using wildcard in the dot notation, this would cover the nested mutations in the `HasMany` case, while ensuring that a comment will be created for the authenticated user or a mutation that uses a list of input types.

Of course, today, I can create something like:

`
input CreateCommentInput {
    post_id: ID!
    title: String!
    body: String!
}

type Mutation @guard {
    createComment(input: CreateCommentInput!): Comment!
        @create
        @inject(context: "user.id", name: "customer_id")
}
`

But, well, this way I'm not using the power of nested mutations.

Maybe I'm wrong, but I think the change needed is just use `data_fill` instead of `Illuminate\Support\Arr::add` in the `Nuwave\Lighthouse\Schema\Directives\InjectDirective` class and rewrite the `addValue` method of `Nuwave\Lighthouse\Execution\Arguments\ArgumentSet` class to behave like `data_fill` as actually it is behaving like `Illuminate\Support\Arr::add`

Or maybe create a new directive and method.

I'm willing to help with a PR when as soon as there is time available.

Lighthouse Version: 4.9
Laravel Version: 6.12.0

### PR DESCRIPTION
**Pull request by andershagbard**
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

Resolves #1206

~I need some help finishing this. It throws a weird error, probably due to how I made the test. If you run `dd($this->toArray)` on line 73 in [`ArgumentSet`](https://github.com/andershagbard/lighthouse/blob/inject-wildcard/src/Execution/Arguments/ArgumentSet.php), you can see it injects the context with the correct value.~

**Changes**

You can now use wildcard in the inject directive.

**Breaking changes**

None I think?

### CODE REVIEW

File: tests/Integration/Schema/Directives/InjectDirectiveTest.php [hunk 1]
```diff
@@ -56,4 +56,61 @@ public function testCreateFromInputObjectWithDeepInjection(): void
...
+            name: String!
+            user: User @belongsTo
+        }
+
+        type User {
+            id: ID
+        }
+
+        type Mutation {
+            createTasks(input: [CreateTaskInput!]! @spread): Task @create @inject(context: "user.id", name: "input.*.user_id")
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/InjectDirectiveTest.php:77]: This is not a valid usage of `@create`, it can only create a single task.

File: src/Execution/Arguments/ArgumentSet.php [hunk 2]
```diff
@@ -61,19 +61,38 @@ public function has(string $key): bool
...
+     * @param  mixed  $value any value to inject
+     * @param  ArgumentSet  $argumentSet
+     */
+    private static function applyValue(array $keys, $value, ArgumentSet $argumentSet): void
+    {
         while (count($keys) > 1) {
             $key = array_shift($keys);
 
+            if ($key === '*') {
+                foreach ($argumentSet as $argument) {
```
Comment (andershagbard) [inline src/Execution/Arguments/ArgumentSet.php:89]: Not sure if we should throw an error with it's not an array here?

File: src/Schema/Directives/InjectDirective.php [hunk 3]
```diff
@@ -28,6 +28,7 @@ public static function definition(): string
   The target name of the argument into which the value is injected.
   You can use dot notation to set the value at arbitrary depth
   within the incoming argument.
+  Asterisks may be used to indicate wildcards.
```
Comment (spawnia) [inline src/Schema/Directives/InjectDirective.php:31]: This does not really explain what it actually does.

```suggestion
  Use an asterisk `*` as a path segment where the value of the argument is a list to be traversed.
```

Also, the definition needs to be copied to the docs - just look for `@inject` in `/docs`. I think we should add an example there.

File: src/Execution/Arguments/ArgumentSet.php (hunk 2 above)
Comment (andershagbard) [reply src/Execution/Arguments/ArgumentSet.php:95]: I've updated it now. Not sure if i used the correct exception.

File: docs/master/api-reference/directives.md [hunk 4]
```diff
@@ -1756,6 +1757,16 @@ type Mutation {
 }
 ```
 
+If you have an array you need to inject a value into, you can use an asterisk `*`. 
+
+```graphql
+type Mutation {
+  createTask(input: CreateTaskInput!): Task
+    @create
+    @inject(context: "user.id", name: "input.*.user_id")
```
Comment (spawnia) [inline docs/master/api-reference/directives.md:1766]: This example does not work.

File: src/Execution/Arguments/ArgumentSet.php (hunk 2 above)
Comment (spawnia) [reply src/Execution/Arguments/ArgumentSet.php:95]: Use `DefinitionException`, in this case the error must be solved by the developer.

File: docs/master/api-reference/directives.md (hunk 4 above)
Comment (andershagbard) [reply docs/master/api-reference/directives.md]: I dont know if the docs has an existing example we can use, but I've updated it to a `updateUser` mutation.

File: docs/master/api-reference/directives.md [hunk 5]
```diff
@@ -1756,6 +1757,16 @@ type Mutation {
...
 ```
 
+If you have an array you need to inject a value into, you can use an asterisk `*`. 
+
+```graphql
+type Mutation {
+  updateUser(input: UpdateUser!): Task
+    @update
+    @inject(context: "user.id", name: "input.tasks.create.*.user_id")
+}
```
[the remark this answers is not in the document]
Comment (andershagbard) [reply docs/master/api-reference/directives.md:1767]: Makes sense. Perhaps we should check whether the fields are mandatory, and if not, just `break` instead of throwing an error?

[the remark this answers is not in the document]
Comment (andershagbard) [reply docs/master/api-reference/directives.md]: I guess the same "issue" could exists with just dot notation. But it would make sense to discard it if an array does not exists at the asterisk segment.

[the remark this answers is not in the document]
Comment (spawnia) [reply docs/master/api-reference/directives.md]: With just dot notation, we force the value at the defined position - and create intermediary inputs if necessary. I guess this could not be desired in every case, but it is how it currently works. At least it is well defined.

In the case of an array, we have the problem that it is less clear what to insert. What number of arguments at the desired position should be added? 1 or another number? I would argue it should always be 0 - that being equivalent to injecting nothing at all.

**Comment by andershagbard** [comment]
I've taken a slightly different approach, just to make it easy. I've implemented `ArrayAccess` and let `data_set` handle the logic.

I've also removed all empty arrays from the values. It caused a `array to string conversion error` later in the code. In this test: `testWillRejectValuesNotPlacedAtArrayWithWildcardInjection`

File: src/Execution/Arguments/ArgumentSet.php [hunk 6]
```diff
@@ -102,4 +102,34 @@ public function argumentsWithUndefined(): array
...
+
+        return $argumentSet->arguments[$offset];
+    }
+
+    public function offsetSet(mixed $offset, mixed $value): void
+    {
+        $argumentSet = $this;
+
+        $argument = new Argument();
+        $argument->value = $value ?: new self();
```
Comment (spawnia) [inline src/Execution/Arguments/ArgumentSet.php:125]: This means `false`, `0` or `''` would not be possible to set

**Review by spawnia — CHANGES_REQUESTED** [review]

File: src/Execution/Arguments/ArgumentSet.php [hunk 7]
```diff
@@ -61,38 +61,38 @@ public function has(string $key): bool
...
+        data_set($argumentSet, $path, $value);
+        self::removeEmptyArrays($argumentSet);
 
         return $this;
     }
 
+    /**
+     * Remove all arguments which is an empty array.
+     */
+    private static function removeEmptyArrays(ArgumentSet $argumentSet): void
```
[the remark this answers is not in the document]
Comment (andershagbard) [reply src/Execution/Arguments/ArgumentSet.php:81]: I think this was a symptom of `data_set` setting object properties. Will remove.

**Comment by andershagbard** [comment]
Not sure how to solve the unit tests.

File: src/Execution/Arguments/Argument.php [hunk 8]
```diff
@@ -85,4 +85,42 @@ protected static function toPlainRecursive($value)
 
         return $value;
     }
+
+    public function offsetExists(mixed $offset): bool
+    {
+        $argument = $this;
```
Comment (spawnia) [inline src/Execution/Arguments/Argument.php:91]: I do not get why you keep aliasing `$this` - why not just it directly?

File: src/Execution/Arguments/Argument.php [hunk 9]
```diff
@@ -85,4 +85,42 @@ protected static function toPlainRecursive($value)
...
+
+        unset($argument[$offset]);
+    }
+
+    public function getIterator(): \ArrayIterator
+    {
+        $value = $this->value;
+
+        return new \ArrayIterator($value);
```
Comment (spawnia) [inline src/Execution/Arguments/Argument.php:124]:
```suggestion
        return new \ArrayIterator($this->value);
```

**Review by spawnia — COMMENTED** [review]
Have you considered adding to `ArgumentSetTest`? It should be pretty straightforward to test the many possible edge cases there and examine exactly what the result will be.
